src/conversions/eyre.rs

```diff
@@ -0,0 +1,158 @@
+#![cfg(feature = "eyre")]
+#![cfg_attr(docsrs, doc(cfg(feature = "eyre")))]
+//! A conversion from [eyre]’s [`Report`] type to [`PyErr`].
+//!
+//! Use of an error handling library like [eyre] is common in application code and when you just
+//! want error handling to be easy. If you are writing a library or you need more control over your
+//! errors you might want to design your own error type instead.
+//!
+//! This implementation always creates a Python [`RuntimeError`]. You might find that you need to
+//! map the error from your Rust code into another Python exception. See [`PyErr::new`] for more
+//! information about that.
+//!
+//! For information about error handling in general, see the [Error handling] chapter of the Rust
+//! book.
+//!
+//! # Setup
+//!
+//! To use this feature, add this to your **`Cargo.toml`**:
+//!
+//! ```toml
+//! [dependencies]
+//! ## change * to the version you want to use, ideally the latest.
+//! eyre = "*"
+// workaround for `extended_key_value_attributes`: https://github.com/rust-lang/rust/issues/82768#issuecomment-803935643
+#![cfg_attr(docsrs, cfg_attr(docsrs, doc = concat!("pyo3 = { version = \"", env!("CARGO_PKG_VERSION"),  "\", features = [\"eyre\"] }")))]
+#![cfg_attr(not(docsrs), doc = "pyo3 = { version = \"*\", features = [\"eyre\"] }")]
+//! ```
+//!
+//! Note that you must use compatible versions of eyre and PyO3.
+//! The required eyre version may vary based on the version of PyO3.
+//!
+//! # Example: Propagating a `PyErr` into [`eyre::Report`]
+//!
+//! ```rust
+//! use pyo3::prelude::*;
+//! use pyo3::wrap_pyfunction;
+//! use std::path::PathBuf;
+//!
+//! // An example function that returns `eyre::Result<...>`.
+//! fn rust_open(filename: PathBuf) -> eyre::Result<Vec<u8>> {
+//!     let data = std::fs::read(filename)?;
+//!     Ok(data)
+//! }
+//!
+//! // A wrapper around a Rust function.
+//! #[pyfunction]
+//! fn py_open(filename: PathBuf) -> PyResult<Vec<u8>> {
+//!     // The `?` ("try") operator performs the conversion
+//!     // into a `PyErr` - if it is an error.
+//!     let data = rust_open(filename)?;
+//!     Ok(data)
+//! }
```

Comment:
I think it should be possible to also just have a `#[pyfunction]` which returns `eyre::Result<Vec<u8>>` - which might be useful to have as an example (as a kind of test)?

Reply:
I thought about doing that but I want to point at the where the conversion happens. A `#[pyfunction]` returning `eyre::Result<Vec<u8>>` has that conversion hidden somewhere in the `pyfunction` macro.

Reply:
That's a fair point. I wonder if your second example "using `eyre` in general" could be tweaked slightly so that the function `decompress` is a `pyfunction` returning `eyre::Result`? The conversion from `PyErr` -> `eyre::Result` can still be there.

Alternatively maybe it's better to keep away from that here, and instead in the guide's `function.md` there is probably scope for an improved "error handling" section which comments on how to use `Result` with error types that convert to `PyErr`. (And can use `Result<_, eyre::Report>` as an example?)